### PR TITLE
[Code standards] Add detail on small PRs and Go packages

### PR DESCRIPTION
**Pull request by lbernick**
This commit updates code standards to include guidance on pull request size, structuring Go packages, and better release notes.

It also removes language around being in a "releasable state".
It's difficult to write a PR that puts us in an unreleasable state, since our CI prevents this,
so guidance that PRs shouldn't put us in an unreleasable state adds confusion.

### CODE REVIEW

File: standards.md
Comment (XinruZhang) [inline standards.md:173]: > Pull requests that make a single, ...

Did we miss "Small" here? like "Small pull requests..."

Comment (XinruZhang) [inline standards.md:173]: > Pull requests that make a single, ...

Did we miss "Small" here? like "Small pull requests..."

File: standards.md
Comment (Yongxuanzhang) [inline standards.md:35]:
```suggestion
  saying they understand what is changing and why:
```
is this a typo?

File: standards.md
Comment (Yongxuanzhang) [inline standards.md:175]:
```suggestion
* Tests and documentation (where applicable) should always be part of the same pull request when introducing a change
```

File: standards.md
Comment (bobcatfish) [inline standards.md:224]: https://go.dev/blog/package-names could be a great link to include here, particularly https://go.dev/blog/package-names#bad-package-names

**Review by bobcatfish — COMMENTED** [review]
Looks great to me! 🎉  Very big fan of making standards like this clear, thanks @lbernick !

/approve

File: standards.md
Comment (lbernick) [reply standards.md]: Added, thanks!

**Comment by lbernick** [comment]
thanks @bobcatfish, just FYI I made a few changes clarifying our guidance around "releasable state" and partial feature development since you approved

File: standards.md
Comment (XinruZhang) [inline standards.md:221]: If we want to move this point under `Go package`, can we rephrase it to make it more focused on how we want to organize "Go packages" instead of how to write tests for go packages?

For example, instead of saying Test exported functionas only, we may want to rephrase it as something like "Every exported functions should be tested"? wdyt?

**Review by vdemeester — APPROVED** [review]
😻

**Review by XinruZhang — APPROVED** [review]
Thanks @lbernick! Looks great! Just a minor comment.

File: standards.md
Comment (lbernick) [reply standards.md]: thanks! reworded a bit

File: standards.md
Comment (bendory) [inline standards.md:180]: suggest: "if you refactor as a prerequisite that will simplify or enable your new changes, commit and merge the refactor before merging your new changes

File: standards.md
Comment (bendory) [inline standards.md:185]: suggest giving an example commit message for this

File: standards.md
Comment (bendory) [inline standards.md:163]: I would suggest keeping this section -- we do want to always be able to cut a release and should avoid merging a PR that puts us in an unreleasable state.

For incremental addition of new features, the incremental changes need to be made in a way that is non-breaking and non-blocking -- such as flag guards and the like.

File: standards.md
Comment (bendory) [inline standards.md:218]: grammar nit: "well-factored" (add a hyphen)

**Review by bendory — APPROVED** [review]

File: standards.md
Comment (lbernick) [reply standards.md:163]: I added a note that new features and api fields should be gated behind feature flags.
I agree we always want to be in an unreleasable state, but having this guidance in the code standards has caused some confusion since it's not clear how we'd get into an "unreleasable" state or what that actually means (thanks @XinruZhang for bringing this up!). I think it may be better to have specific guidance on what types of PRs to avoid. WDYT?

File: standards.md
Comment (XinruZhang) [inline standards.md:70]: Thanks Lee! I love this new instruction, we should really only announce the a new feature in release note when it's done. (Nit: do we want to call out the API Change case? even though it's user facing, we shouldn;t mention it in the release note unless the implementation is in place.)

Comment (lbernick) [reply standards.md:70]: I did actually :) take a look at the example under "incremental feature development"

Comment (XinruZhang) [reply standards.md:70]: Ahh that looks great. Thanks Lee! Would it be helpful if we link that section ("Incremental Feature Development") here as a reference?

Comment (lbernick) [reply standards.md:70]: done!

File: standards.md
Comment (bendory) [reply standards.md:198]: Agree that it's a bit overspecified -- but I think it's helpful to overspecify for people new to this "small PR" principle in development.

/LGTM

**Comment by lbernick** [comment]
@bobcatfish @vdemeester would either of you mind giving a LGTM please?

File: standards.md
Comment (afrittoli) [reply standards.md:163]: I'm not sure about removing this. 

I think PR authors should ask themselves the question "if this was the last PR before the release, would we produce a usable release?". This is important to make sure that our nightly builds are usable as well as it helps very much in reducing the complexity the release manager may have to deal with.

Of course, CI takes a big role in ensuring that the code can be compiled and is generally safe and functionally sound. Feature flags help with incomplete features, but also documentation should be written so that it does not mislead users. We can try to make a complete list of all the things to consider, but I would prefer we still ask authors to ask themselves the question about the "releasable" state.

**Review by afrittoli — COMMENTED** [review]
Thank a lot @lbernick - this looks good - only I disagree about removing the part about "releasable" state completely.

File: standards.md
Comment (bendory) [reply standards.md:163]: A few affirmative guidelines and / or antipatterns that I can think of:

- No PR should require concurrent release of code in another repository
- No PR should require that a release be held pending some later PR
- No PR should include a breaking change without either (1) a controlling feature flag or (2) a release note that explains the intentionality of the breaking change
- A PR that intentionally changes functionality should only be merged along with compatible, passing tests in the same PR

Comment (lbernick) [reply standards.md:163]: Thanks for the discussion here! I brought back the note about "unreleasable state" with a bit more explanation. @bendory I included all your examples but changed the third to "PRs must adhere to the project's API stability policy." I think it would be good as a follow up to move most of the contents of https://github.com/tektoncd/pipeline/blob/main/api_compatibility_policy.md that aren't specific to pipelines into the community repo so that all projects can link them.

**Review by afrittoli — COMMENTED** [review]
Thanks for the updates @lbernick - it looks good to me now!
/approve

File: standards.md
Comment (bendory) [reply standards.md:163]: /LGTM

modulo one minor nit in a separate comment

**Review by bendory — APPROVED** [review]

**Review by jerop — APPROVED** [review]
/lgtm

**Pull request merged**
